components/ws-proxy/pkg/sshproxy/server.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package sshproxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	supervisor "github.com/gitpod-io/gitpod/supervisor/api"
+	p "github.com/gitpod-io/gitpod/ws-proxy/pkg/proxy"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/xerrors"
+	"google.golang.org/grpc"
+)
+
+type Session struct {
+	Conn                *ssh.ServerConn
+	WorkspaceId         string
+	PublicKey           ssh.PublicKey
+	WorkspaceIp         string
+	WorkspacePrivateKey ssh.Signer
+}
+
+type Server struct {
+	ConnectionTimeout     time.Duration
+	sshConfig             *ssh.ServerConfig
+	workspaceInfoProvider p.WorkspaceInfoProvider
+}
+
+func (s *Server) HandleConn(c net.Conn) {
+	sshConn, chans, reqs, err := ssh.NewServerConn(c, s.sshConfig)
+	if err != nil {
+		c.Close()
+		return
+	}
+	defer sshConn.Close()
+
+	if sshConn.Permissions == nil || sshConn.Permissions.Extensions == nil || sshConn.Permissions.Extensions["workspaceId"] == "" {
+		return
+	}
+	workspaceId := sshConn.Permissions.Extensions["workspaceId"]
+	wsInfo := s.workspaceInfoProvider.WorkspaceInfo(workspaceId)
+	if wsInfo == nil {
+		return
+	}
+	ctx, _ := context.WithTimeout(context.Background(), time.Second*5)
+	key, err := s.GetWorkspaceSSHKey(ctx, wsInfo.IPAddress)
+	if err != nil {
+		return
+	}
+	session := &Session{
+		Conn:                sshConn,
+		WorkspaceId:         workspaceId,
+		WorkspacePrivateKey: key,
+		WorkspaceIp:         wsInfo.IPAddress + ":23001",
+	}
+
+	go func() {
+		for req := range reqs {
+			switch req.Type {
+			case "keepalive@openssh.com":
+				if req.WantReply {
+					req.Reply(true, []byte{})
+				}
+			default:
+				req.Reply(false, []byte{})
+			}
+		}
+	}()
+
+	for newChannel := range chans {
+		switch newChannel.ChannelType() {
+		case "session":
+			go s.SessionForward(session, newChannel)
+		case "direct-tcpip":
+			go s.ChannelForward(session, newChannel)
+		case "tcpip-forward":
+			newChannel.Reject(ssh.UnknownChannelType, "Gitpod SSH Gateway cannot remote forward ports")
+		default:
+			newChannel.Reject(ssh.UnknownChannelType, fmt.Sprintf("Gitpod SSH Gateway cannot handle %s channel types", newChannel.ChannelType()))
```

Comment:
it may be worth to log that we know

Reply:
this may spam the logs. Consider introducing a counter metric instead